File: mobile/src/main/java/com/ataulm/chunks/ChunksActions.java
package com.ataulm.chunks;

import com.ataulm.Optional;
import com.google.auto.value.AutoValue;
import com.novoda.accessibility.Action;
import com.novoda.accessibility.Actions;

import java.util.ArrayList;

@AutoValue
abstract class
ChunksActions {

    public static ChunksActions create(Items items, Day day, Item item, ItemUserInteractions userInteractions) {
        Optional<Action> markComplete = createMarkCompleteAction(item, userInteractions);
        Optional<Action> markNotComplete = createMarkNotCompleteAction(item, userInteractions);
        Action edit = createEditAction(item, userInteractions);
        Optional<Action> transitionToToday = createTransitionToTodayAction(day, item, userInteractions);
        Optional<Action> transitionToTomorrow = createTransitionToTomorrowAction(day, item, userInteractions);
        Optional<Action> transitionToLater = createTransitionToLaterAction(day, item, userInteractions);
        Optional<Action> moveUp = createMoveUpAction(items, item, userInteractions);
        Optional<Action> moveDown = createMoveDownAction(items, item, userInteractions);
        Action delete = createDeleteAction(item, userInteractions);

        Actions actions = collate(markComplete, markNotComplete, edit, transitionToToday, transitionToTomorrow, transitionToLater, moveUp, moveDown, delete);
        return new AutoValue_ChunksActions(actions, markComplete, markNotComplete, edit, transitionToToday, transitionToTomorrow, transitionToLater, moveUp, moveDown, delete, day);
    }

    private static Optional<Action> createMarkCompleteAction(final Item item, final ItemUserInteractions userInteractions) {
        if (item.isCompleted()) {
            return Optional.absent();
        }
        return Optional.of(new Action(R.id.action_mark_complete, R.string.action_mark_complete, new Runnable() {
            @Override
            public void run() {
                userInteractions.onUserMarkComplete(item);
            }
        }));
    }

    private static Optional<Action> createMarkNotCompleteAction(final Item item, final ItemUserInteractions userInteractions) {
        if (!item.isCompleted()) {
            return Optional.absent();
        }
        return Optional.of(new Action(R.id.action_mark_not_complete, R.string.action_mark_not_complete, new Runnable() {
            @Override
            public void run() {
                userInteractions.onUserMarkNotComplete(item);
            }
        }));
    }

    private static Action createEditAction(final Item item, final ItemUserInteractions userInteractions) {
        return new Action(R.id.action_edit, R.string.action_edit, new Runnable() {
            @Override
            public void run() {
                userInteractions.onUserEdit(item);
            }
        });
    }

    private static Optional<Action> createTransitionToTodayAction(Day day, final Item item, final ItemUserInteractions userInteractions) {
        if (day == Day.TODAY) {
            return Optional.absent();
        }
        return Optional.of(new Action(R.id.action_move_to_today, R.string.action_move_to_today, new Runnable() {
            @Override
            public void run() {
                userInteractions.onUserTransitionItem(item, Day.TODAY);
            }
        }));
    }

    private static Optional<Action> createTransitionToTomorrowAction(Day day, final Item item, final ItemUserInteractions userInteractions) {
        if (day == Day.TOMORROW) {
            return Optional.absent();
        }
        return Optional.of(new Action(R.id.action_move_to_tomorrow, R.string.action_move_to_tomorrow, new Runnable() {
            @Override
            public void run() {
                userInteractions.onUserTransitionItem(item, Day.TOMORROW);
            }
        }));
    }

    private static Optional<Action> createTransitionToLaterAction(Day day, final Item item, final ItemUserInteractions userInteractions) {
        if (day == Day.SOMETIME) {
            return Optional.absent();
        }
        return Optional.of(new Action(R.id.action_move_to_later, R.string.action_move_to_later, new Runnable() {
            @Override
            public void run() {
                userInteractions.onUserTransitionItem(item, Day.SOMETIME);
            }
        }));
    }

    private static Optional<Action> createMoveUpAction(final Items items, final Item item, final ItemUserInteractions userInteractions) {
        final int originalEntryPosition = items.entries().indexOf(item);
        if (originalEntryPosition == 0) {
            return Optional.absent();
        }
        return Optional.of(new Action(R.id.action_move_up, R.string.action_move_up, new Runnable() {
            @Override
            public void run() {
                userInteractions.onUserMove(item, originalEntryPosition - 1);
            }
        }));
    }

    private static Optional<Action> createMoveDownAction(final Items items, final Item item, final ItemUserInteractions userInteractions) {
        final int originalEntryPosition = items.entries().indexOf(item);
        if (originalEntryPosition == items.size() - 1) {
            return Optional.absent();
        }
        return Optional.of(new Action(R.id.action_move_down, R.string.action_move_down, new Runnable() {
            @Override
            public void run() {
                userInteractions.onUserMove(item, originalEntryPosition + 1);
            }
        }));
    }

    private static Action createDeleteAction(final Item item, final ItemUserInteractions userInteractions) {
        return new Action(R.id.action_delete, R.string.action_delete, new Runnable() {
            @Override
            public void run() {
                userInteractions.onUserRemove(item);
            }
        });
    }

    private static Actions collate(
            Optional<Action> markComplete,
            Optional<Action> markNotComplete,
            Action edit,
            Optional<Action> transitionToToday,
            Optional<Action> transitionToTomorrow,
            Optional<Action> transitionToLater,
            Optional<Action> moveUp,
            Optional<Action> moveDown,
            Action delete
    ) {
        ArrayList<Action> actions = new ArrayList<>();
        if (markComplete.isPresent()) {
            actions.add(markComplete.get());
        }
        if (markNotComplete.isPresent()) {
            actions.add(markNotComplete.get());
        }
        actions.add(edit);
        if (transitionToToday.isPresent()) {
            actions.add(transitionToToday.get());
        }
        if (transitionToTomorrow.isPresent()) {
            actions.add(transitionToTomorrow.get());
        }
        if (transitionToLater.isPresent()) {
            actions.add(transitionToLater.get());
        }
        if (moveUp.isPresent()) {
            actions.add(moveUp.get());
        }
        if (moveDown.isPresent()) {
            actions.add(moveDown.get());
        }
        actions.add(delete);
        return new Actions(actions);
    }

    abstract Actions actions();

    abstract Optional<Action> markComplete();

    abstract Optional<Action> markNotComplete();

    abstract Action edit();

    abstract Optional<Action> transitionToToday();

    abstract Optional<Action> transitionToTomorrow();

    abstract Optional<Action> transitionToLater();

    abstract Optional<Action> moveUp();

    abstract Optional<Action> moveDown();

    abstract Action delete();

    abstract Day currentDay();

    Optional<Action> transitionToPreviousDay() {
        Day day = currentDay();
        switch (day) {
            case TODAY:
                return Optional.absent();
            case TOMORROW:
                return transitionToToday();
            case SOMETIME:
                return transitionToTomorrow();
            default:
                throw new IllegalArgumentException("unknown day: " + day);
        }
    }

    Optional<Action> transitionToNextDay() {
        Day day = currentDay();
        switch (day) {
            case TODAY:
                return transitionToTomorrow();
            case TOMORROW:
                return transitionToLater();
            case SOMETIME:
                return Optional.absent();
            default:
                throw new IllegalArgumentException("unknown day: " + day);
        }
    }

}

File: mobile/src/main/java/com/ataulm/chunks/ChunksActions.kt
package com.ataulm.chunks

import com.novoda.accessibility.Action
import com.novoda.accessibility.Actions
import java.util.*

data class ChunksActions(
        val actions: Actions,
        val markComplete: Action?,
        val markNotComplete: Action?,
        val edit: Action,
        val transitionToToday: Action?,
        val transitionToTomorrow: Action?,
        val transitionToLater: Action?,
        val moveUp: Action?,
        val moveDown: Action?,
        val delete: Action,
        val currentDay: Day
) {

    fun transitionToPreviousDay(): Action? {
        val day = currentDay
        return when (day) {
            Day.TODAY -> null
            Day.TOMORROW -> transitionToToday
            Day.SOMETIME -> transitionToTomorrow
        }
    }

    fun transitionToNextDay(): Action? {
        val day = currentDay
        return when (day) {
            Day.TODAY -> transitionToTomorrow
            Day.TOMORROW -> transitionToLater
            Day.SOMETIME -> null
        }
    }

    companion object {

        fun create(items: Items, day: Day, item: Item, userInteractions: ItemUserInteractions): ChunksActions {
            val markComplete = createMarkCompleteAction(item, userInteractions)
            val markNotComplete = createMarkNotCompleteAction(item, userInteractions)
            val edit = createEditAction(item, userInteractions)
            val transitionToToday = createTransitionToTodayAction(day, item, userInteractions)
            val transitionToTomorrow = createTransitionToTomorrowAction(day, item, userInteractions)
            val transitionToLater = createTransitionToLaterAction(day, item, userInteractions)
            val moveUp = createMoveUpAction(items, item, userInteractions)
            val moveDown = createMoveDownAction(items, item, userInteractions)
            val delete = createDeleteAction(item, userInteractions)

            val actions = collate(markComplete, markNotComplete, edit, transitionToToday, transitionToTomorrow, transitionToLater, moveUp, moveDown, delete)
            return ChunksActions(actions, markComplete, markNotComplete, edit, transitionToToday, transitionToTomorrow, transitionToLater, moveUp, moveDown, delete, day)
        }

        private fun createMarkCompleteAction(item: Item, userInteractions: ItemUserInteractions): Action? {
            return if (item.isCompleted) {
                null
            } else Action(R.id.action_mark_complete, R.string.action_mark_complete, Runnable { userInteractions.onUserMarkComplete(item) })
        }

        private fun createMarkNotCompleteAction(item: Item, userInteractions: ItemUserInteractions): Action? {
            return if (!item.isCompleted) {
                null
            } else Action(R.id.action_mark_not_complete, R.string.action_mark_not_complete, Runnable { userInteractions.onUserMarkNotComplete(item) })
        }

        private fun createEditAction(item: Item, userInteractions: ItemUserInteractions): Action {
            return Action(R.id.action_edit, R.string.action_edit, Runnable { userInteractions.onUserEdit(item) })
        }

        private fun createTransitionToTodayAction(day: Day, item: Item, userInteractions: ItemUserInteractions): Action? {
            return if (day == Day.TODAY) {
                null
            } else Action(R.id.action_move_to_today, R.string.action_move_to_today, Runnable { userInteractions.onUserTransitionItem(item, Day.TODAY) })
        }

        private fun createTransitionToTomorrowAction(day: Day, item: Item, userInteractions: ItemUserInteractions): Action? {
            return if (day == Day.TOMORROW) {
                null
            } else Action(R.id.action_move_to_tomorrow, R.string.action_move_to_tomorrow, Runnable { userInteractions.onUserTransitionItem(item, Day.TOMORROW) })
        }

        private fun createTransitionToLaterAction(day: Day, item: Item, userInteractions: ItemUserInteractions): Action? {
            return if (day == Day.SOMETIME) {
                null
            } else Action(R.id.action_move_to_later, R.string.action_move_to_later, Runnable { userInteractions.onUserTransitionItem(item, Day.SOMETIME) })
        }

        private fun createMoveUpAction(items: Items, item: Item, userInteractions: ItemUserInteractions): Action? {
            val originalEntryPosition = items.entries.indexOf(item)
            return if (originalEntryPosition == 0) {
                null
            } else Action(R.id.action_move_up, R.string.action_move_up, Runnable { userInteractions.onUserMove(item, originalEntryPosition - 1) })
        }

        private fun createMoveDownAction(items: Items, item: Item, userInteractions: ItemUserInteractions): Action? {
            val originalEntryPosition = items.entries.indexOf(item)
            return if (originalEntryPosition == items.size() - 1) {
                null
            } else Action(R.id.action_move_down, R.string.action_move_down, Runnable { userInteractions.onUserMove(item, originalEntryPosition + 1) })
        }

        private fun createDeleteAction(item: Item, userInteractions: ItemUserInteractions): Action {
            return Action(R.id.action_delete, R.string.action_delete, Runnable { userInteractions.onUserRemove(item) })
        }

        private fun collate(
                markComplete: Action?,
                markNotComplete: Action?,
                edit: Action,
                transitionToToday: Action?,
                transitionToTomorrow: Action?,
                transitionToLater: Action?,
                moveUp: Action?,
                moveDown: Action?,
                delete: Action
        ): Actions {
            val actions = ArrayList<Action>()
            markComplete?.let { actions.add(it) }
            markNotComplete?.let { actions.add(it) }
            actions.add(edit)
            transitionToToday?.let { actions.add(it) }
            transitionToTomorrow?.let { actions.add(it) }
            transitionToLater?.let { actions.add(it) }
            moveUp?.let { actions.add(it) }
            moveDown?.let { actions.add(it) }
            actions.add(delete)
            return Actions(actions)
        }
    }

}
